File: mobile_deployment/mobile_deployment/examples/android/GlucosePredictor.java
package com.example.glucoseprediction;

import android.content.res.AssetFileDescriptor;
import android.content.res.AssetManager;
import org.tensorflow.lite.Interpreter;
import org.tensorflow.lite.Tensor;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;
import java.util.HashMap;
import java.util.Map;

/**
 * 血糖预测模型 - Android TensorFlow Lite 集成示例
 * 基于LSTM + Cross-Attention架构的血糖水平预测模型
 *
 * 功能：
 * - 预测未来15、30、45、60分钟的血糖水平
 * - 输入：10个时间步的历史数据(51特征) + 静态患者特征(30特征)
 * - 输出：4个时间点的血糖预测值
 */
public class GlucosePredictor {
    private static final String MODEL_FILE = "glucose_predictor.tflite";
    private static final int TIME_SERIES_LENGTH = 10;
    private static final int TIME_SERIES_FEATURES = 51;
    private static final int STATIC_FEATURES = 30;
    private static final int OUTPUT_LENGTH = 4;

    private Interpreter interpreter;
    private boolean isInitialized = false;

    // 预测时间点（分钟）
    private static final int[] PREDICTION_HORIZONS = {15, 30, 45, 60};

    /**
     * 初始化模型
     * @param assetManager Android资源管理器
     */
    public void initialize(AssetManager assetManager) throws IOException {
        if (isInitialized) {
            return;
        }

        // 加载TFLite模型
        MappedByteBuffer modelBuffer = loadModelFile(assetManager, MODEL_FILE);

        // 创建解释器选项
        Interpreter.Options options = new Interpreter.Options();
        options.setNumThreads(4); // 使用4个线程进行推理

        // 创建解释器
        this.interpreter = new Interpreter(modelBuffer, options);
        this.isInitialized = true;

        // 输出模型信息
        logModelInfo();
    }

    /**
     * 预测血糖水平
     * @param timeSeriesData 10个时间步的历史数据 [10][51]
     * @param staticData 静态患者特征 [30]
     * @return 预测结果 Map<时间点(分钟), 血糖预测值>
     */
    public Map<Integer, Float> predictGlucose(float[][] timeSeriesData, float[] staticData) {
        if (!isInitialized) {
            throw new IllegalStateException("模型未初始化，请先调用initialize()");
        }

        // 验证输入数据维度
        validateInputData(timeSeriesData, staticData);

        // 准备输入缓冲区
        ByteBuffer timeSeriesBuffer = ByteBuffer.allocateDirect(
            TIME_SERIES_LENGTH * TIME_SERIES_FEATURES * Float.BYTES);
        timeSeriesBuffer.order(ByteOrder.nativeOrder());

        ByteBuffer staticBuffer = ByteBuffer.allocateDirect(
            STATIC_FEATURES * Float.BYTES);
        staticBuffer.order(ByteOrder.nativeOrder());

        // 填充时序数据
        for (int i = 0; i < TIME_SERIES_LENGTH; i++) {
            for (int j = 0; j < TIME_SERIES_FEATURES; j++) {
                timeSeriesBuffer.putFloat(timeSeriesData[i][j]);
            }
        }
        timeSeriesBuffer.rewind();

        // 填充静态数据
        for (int i = 0; i < STATIC_FEATURES; i++) {
            staticBuffer.putFloat(staticData[i]);
        }
        staticBuffer.rewind();

        // 准备输出缓冲区
        float[][] outputData = new float[1][OUTPUT_LENGTH];

        // 执行推理
        Object[] inputs = {timeSeriesBuffer, staticBuffer};
        Map<Integer, Object> outputs = new HashMap<>();
        outputs.put(0, outputData);

        interpreter.runForMultipleInputsOutputs(inputs, outputs);

        // 处理输出结果
        Map<Integer, Float> predictions = new HashMap<>();
        float[] rawPredictions = outputData[0];

        for (int i = 0; i < OUTPUT_LENGTH; i++) {
            predictions.put(PREDICTION_HORIZONS[i], rawPredictions[i]);
        }

        return predictions;
    }

    /**
     * 批量预测（用于性能测试）
     * @param batchData 批量输入数据
     * @return 批量预测结果
     */
    public Map<Integer, Float>[] predictBatch(float[][][] batchTimeSeriesData, float[][] batchStaticData) {
        if (!isInitialized) {
            throw new IllegalStateException("模型未初始化，请先调用initialize()");
        }

        int batchSize = batchTimeSeriesData.length;
        Map<Integer, Float>[] batchResults = new HashMap[batchSize];

        for (int i = 0; i < batchSize; i++) {
            batchResults[i] = predictGlucose(batchTimeSeriesData[i], batchStaticData[i]);
        }

        return batchResults;
    }

    /**
     * 获取模型信息
     */
    public String getModelInfo() {
        if (!isInitialized) {
            return "模型未初始化";
        }

        StringBuilder info = new StringBuilder();
        info.append("血糖预测模型信息:\\n");
        info.append("- 模型类型: LSTM + Cross-Attention\\n");
        info.append("- 输入: 时序数据[10,51] + 静态特征[30]\\n");
        info.append("- 输出: 血糖预测[4] (15/30/45/60分钟)\\n");

        // 获取输入输出张量信息
        Tensor[] inputTensors = interpreter.getInputTensors();
        Tensor[] outputTensors = interpreter.getOutputTensors();

        info.append(String.format("- 输入数量: %d\\n", inputTensors.length));
        for (int i = 0; i < inputTensors.length; i++) {
            Tensor tensor = inputTensors[i];
            info.append(String.format("  输入%d: %s (%s)\\n",
                i + 1,
                shapeToString(tensor.shape()),
                tensor.dataType()));
        }

        info.append(String.format("- 输出数量: %d\\n", outputTensors.length));
        for (int i = 0; i < outputTensors.length; i++) {
            Tensor tensor = outputTensors[i];
            info.append(String.format("  输出%d: %s (%s)\\n",
                i + 1,
                shapeToString(tensor.shape()),
                tensor.dataType()));
        }

        return info.toString();
    }

    /**
     * 释放资源
     */
    public void close() {
        if (interpreter != null) {
            interpreter.close();
            isInitialized = false;
        }
    }

    // ========== 私有方法 ==========

    /**
     * 从Assets加载模型文件
     */
    private MappedByteBuffer loadModelFile(AssetManager assetManager, String modelPath) throws IOException {
        AssetFileDescriptor fileDescriptor = assetManager.openFd(modelPath);
        FileInputStream inputStream = new FileInputStream(fileDescriptor.getFileDescriptor());
        FileChannel fileChannel = inputStream.getChannel();
        long startOffset = fileDescriptor.getStartOffset();
        long declaredLength = fileDescriptor.getDeclaredLength();

        return fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength);
    }

    /**
     * 验证输入数据维度
     */
    private void validateInputData(float[][] timeSeriesData, float[] staticData) {
        if (timeSeriesData.length != TIME_SERIES_LENGTH) {
            throw new IllegalArgumentException(
                String.format("时序数据长度错误，期望%d，实际%d",
                    TIME_SERIES_LENGTH, timeSeriesData.length));
        }

        for (int i = 0; i < timeSeriesData.length; i++) {
            if (timeSeriesData[i].length != TIME_SERIES_FEATURES) {
                throw new IllegalArgumentException(
                    String.format("时序数据特征数错误，期望%d，实际%d",
                        TIME_SERIES_FEATURES, timeSeriesData[i].length));
            }
        }

        if (staticData.length != STATIC_FEATURES) {
            throw new IllegalArgumentException(
                String.format("静态特征数错误，期望%d，实际%d",
                    STATIC_FEATURES, staticData.length));
        }
    }

    /**
     * 将张量形状转换为字符串
     */
    private String shapeToString(int[] shape) {
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        for (int i = 0; i < shape.length; i++) {
            if (i > 0) sb.append(", ");
            sb.append(shape[i]);
        }
        sb.append("]");
        return sb.toString();
    }

    /**
     * 输出模型信息
     */
    private void logModelInfo() {
        System.out.println("血糖预测模型初始化成功");
        System.out.println(getModelInfo());
    }

    // ========== 使用示例 ==========

    /**
     * 使用示例代码
     */
    public static void usageExample() {
        // 初始化预测器
        GlucosePredictor predictor = new GlucosePredictor();
        try {
            // AssetManager assetManager = context.getAssets();
            // predictor.initialize(assetManager);

            // 准备测试数据
            float[][] timeSeriesData = new float[10][51];
            float[] staticData = new float[30];

            // 填充数据（这里使用随机数据作为示例）
            for (int i = 0; i < 10; i++) {
                for (int j = 0; j < 51; j++) {
                    timeSeriesData[i][j] = (float) Math.random();
                }
            }

            for (int i = 0; i < 30; i++) {
                staticData[i] = (float) Math.random();
            }

            // 执行预测
            Map<Integer, Float> predictions = predictor.predictGlucose(timeSeriesData, staticData);

            // 输出结果
            System.out.println("血糖预测结果:");
            for (Map.Entry<Integer, Float> entry : predictions.entrySet()) {
                System.out.printf("%d分钟后: %.2f mg/dL\\n", entry.getKey(), entry.getValue());
            }

        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            predictor.close();
        }
    }
}

File: android_glucose_app/app/src/main/java/com/glucosepredictor/GlucosePredictor.kt
package com.glucosepredictor

import android.content.Context
import android.util.Log
import org.tensorflow.lite.Interpreter
import org.json.JSONObject
import java.io.FileInputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel

/**
 * 血糖预测器 - 基于TFLite模型
 * 对应Python demo3_tflite_model.py的逻辑
 */
class GlucosePredictor(context: Context) {

    private var interpreter: Interpreter? = null
    private val tag = "GlucosePredictor"

    // 模型输入输出规格
    private val timeSeriesShape = intArrayOf(1, 10, 51)  // (batch, timesteps, features)
    private val staticShape = intArrayOf(1, 30)          // (batch, features)
    private val outputShape = intArrayOf(1, 4)           // (batch, predictions)

    // 标准化参数 (需要从训练数据计算，这里使用示例值)
    private val tsXMean = FloatArray(51) // 时序特征均值
    private val tsXStd = FloatArray(51)  // 时序特征标准差
    private val staticMean = FloatArray(30) // 静态特征均值
    private val staticStd = FloatArray(30)  // 静态特征标准差
    private val yMean = FloatArray(4)    // 目标均值
    private val yStd = FloatArray(4)     // 目标标准差

    init {
        try {
            // 加载TFLite模型
            val modelFile = loadModelFile(context, "glucose_predictor.tflite")
            val options = Interpreter.Options().apply {
                setNumThreads(4)
                setUseXNNPACK(true)
            }
            interpreter = Interpreter(modelFile, options)

            // 从JSON文件加载标准化参数
            loadScalerParams(context)

            Log.d(tag, "✓ 模型加载成功")
        } catch (e: Exception) {
            Log.e(tag, "模型加载失败: ${e.message}")
            e.printStackTrace()
        }
    }

    /**
     * 加载TFLite模型文件
     */
    private fun loadModelFile(context: Context, filename: String): MappedByteBuffer {
        val fileDescriptor = context.assets.openFd(filename)
        val inputStream = FileInputStream(fileDescriptor.fileDescriptor)
        val fileChannel = inputStream.channel
        val startOffset = fileDescriptor.startOffset
        val declaredLength = fileDescriptor.declaredLength
        return fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength)
    }

    /**
     * 从JSON文件加载标准化参数
     */
    private fun loadScalerParams(context: Context) {
        try {
            // 读取JSON文件
            val jsonString = context.assets.open("scaler_params.json").bufferedReader().use { it.readText() }
            val json = JSONObject(jsonString)

            // 加载时序特征参数
            val tsJson = json.getJSONObject("time_series")
            val tsMeanArray = tsJson.getJSONArray("mean")
            val tsStdArray = tsJson.getJSONArray("std")
            for (i in 0 until tsMeanArray.length()) {
                tsXMean[i] = tsMeanArray.getDouble(i).toFloat()
                tsXStd[i] = tsStdArray.getDouble(i).toFloat()
            }

            // 加载静态特征参数
            val staticJson = json.getJSONObject("static")
            val staticMeanArray = staticJson.getJSONArray("mean")
            val staticStdArray = staticJson.getJSONArray("std")
            for (i in 0 until staticMeanArray.length()) {
                staticMean[i] = staticMeanArray.getDouble(i).toFloat()
                staticStd[i] = staticStdArray.getDouble(i).toFloat()
            }

            // 加载目标参数
            val targetJson = json.getJSONObject("target")
            val targetMeanArray = targetJson.getJSONArray("mean")
            val targetStdArray = targetJson.getJSONArray("std")
            for (i in 0 until targetMeanArray.length()) {
                yMean[i] = targetMeanArray.getDouble(i).toFloat()
                yStd[i] = targetStdArray.getDouble(i).toFloat()
            }

            Log.d(tag, "✓ Scaler参数加载成功")
            Log.d(tag, "  时序: mean[0]=${tsXMean[0]}, std[0]=${tsXStd[0]}")
            Log.d(tag, "  目标: mean[0]=${yMean[0]}, std[0]=${yStd[0]}")

        } catch (e: Exception) {
            Log.e(tag, "Scaler参数加载失败: ${e.message}")
            e.printStackTrace()

            // 使用默认值作为后备
            for (i in tsXMean.indices) {
                tsXMean[i] = 0f
                tsXStd[i] = 1f
            }
            for (i in staticMean.indices) {
                staticMean[i] = 0f
                staticStd[i] = 1f
            }
            for (i in yMean.indices) {
                yMean[i] = 130f
                yStd[i] = 30f
            }
        }
    }

    /**
     * 标准化时序数据
     */
    private fun standardizeTimeSeries(data: FloatArray): FloatArray {
        val result = FloatArray(data.size)
        for (i in data.indices) {
            val featureIdx = i % 51
            result[i] = if (tsXStd[featureIdx] != 0f) {
                (data[i] - tsXMean[featureIdx]) / tsXStd[featureIdx]
            } else {
                0f
            }
        }
        return result
    }

    /**
     * 标准化静态数据
     */
    private fun standardizeStatic(data: FloatArray): FloatArray {
        val result = FloatArray(data.size)
        for (i in data.indices) {
            result[i] = if (staticStd[i] != 0f) {
                (data[i] - staticMean[i]) / staticStd[i]
            } else {
                0f
            }
        }
        return result
    }

    /**
     * 反标准化预测结果
     */
    private fun inverseStandardizePredictions(data: FloatArray): FloatArray {
        val result = FloatArray(data.size)
        for (i in data.indices) {
            result[i] = data[i] * yStd[i] + yMean[i]
        }
        return result
    }

    /**
     * 执行预测
     */
    fun predict(timeSeriesData: FloatArray, staticData: FloatArray): FloatArray? {
        try {
            // 标准化输入
            val normalizedTS = standardizeTimeSeries(timeSeriesData)
            val normalizedStatic = standardizeStatic(staticData)

            // 准备输入ByteBuffer
            val tsBuffer = ByteBuffer.allocateDirect(4 * timeSeriesShape[1] * timeSeriesShape[2])
                .order(ByteOrder.nativeOrder())
            normalizedTS.forEach { tsBuffer.putFloat(it) }

            val staticBuffer = ByteBuffer.allocateDirect(4 * staticShape[1])
                .order(ByteOrder.nativeOrder())
            normalizedStatic.forEach { staticBuffer.putFloat(it) }

            // 准备输出
            val outputBuffer = ByteBuffer.allocateDirect(4 * outputShape[1])
                .order(ByteOrder.nativeOrder())

            // 执行推理
            val inputs = arrayOf<Any>(tsBuffer, staticBuffer)
            val outputs = mapOf(0 to outputBuffer)

            interpreter?.runForMultipleInputsOutputs(inputs, outputs)

            // 解析输出
            outputBuffer.rewind()
            val predictions = FloatArray(4)
            outputBuffer.asFloatBuffer().get(predictions)

            // 反标准化
            return inverseStandardizePredictions(predictions)

        } catch (e: Exception) {
            Log.e(tag, "预测失败: ${e.message}")
            e.printStackTrace()
            return null
        }
    }

    /**
     * 预测所有场景
     * 基于demo3_tflite_model.py的时间模拟策略
     */
    fun predictAllScenarios(
        timeSeriesData: FloatArray,
        staticData: FloatArray
    ): PredictionResult {
        // 场景1: 完整输入
        val predFull = predict(timeSeriesData, staticData) ?: floatArrayOf(0f, 0f, 0f, 0f)

        // 场景2: 无患者信息
        val staticZero = FloatArray(30) { 0f }
        val predNoStatic = predict(timeSeriesData, staticZero) ?: floatArrayOf(0f, 0f, 0f, 0f)

        // 场景3: 低热量进餐 (30分钟前轻食)
        // Dietary intake索引=1，在倒数第3个时间步（索引-3）设置为1
        val tsLowMeal = timeSeriesData.copyOf()
        val dietaryIdxLow = tsLowMeal.size - 3 * 51 + 1  // 倒数第3个时间步的Dietary intake
        tsLowMeal[dietaryIdxLow] = 1.0f
        val predLowMeal = predict(tsLowMeal, staticData) ?: floatArrayOf(0f, 0f, 0f, 0f)

        // 场景4: 中热量进餐 (15分钟前正常进餐)
        // Dietary intake索引=1，在倒数第2个时间步（索引-2）设置为1
        val tsMidMeal = timeSeriesData.copyOf()
        val dietaryIdxMid = tsMidMeal.size - 2 * 51 + 1  // 倒数第2个时间步的Dietary intake
        tsMidMeal[dietaryIdxMid] = 1.0f
        val predMidMeal = predict(tsMidMeal, staticData) ?: floatArrayOf(0f, 0f, 0f, 0f)

        // 场景5: 高热量进餐 (持续进餐，30-15分钟前)
        // 在倒数第3和倒数第2个时间步都设置Dietary intake=1
        val tsHighMeal = timeSeriesData.copyOf()
        tsHighMeal[dietaryIdxLow] = 1.0f  // 30分钟前开始进餐
        tsHighMeal[dietaryIdxMid] = 1.0f  // 15分钟前仍在进餐
        val predHighMeal = predict(tsHighMeal, staticData) ?: floatArrayOf(0f, 0f, 0f, 0f)

        return PredictionResult(
            fullInput = predFull,
            noPatientInfo = predNoStatic,
            lowCalorieMeal = predLowMeal,
            mediumCalorieMeal = predMidMeal,
            highCalorieMeal = predHighMeal
        )
    }

    fun close() {
        interpreter?.close()
        interpreter = null
    }
}

/**
 * 预测结果数据类
 */
data class PredictionResult(
    val fullInput: FloatArray,           // 完整输入预测
    val noPatientInfo: FloatArray,       // 无患者信息预测
    val lowCalorieMeal: FloatArray,      // 低热量进餐预测 (30分钟前轻食)
    val mediumCalorieMeal: FloatArray,   // 中热量进餐预测 (15分钟前正常进餐)
    val highCalorieMeal: FloatArray      // 高热量进餐预测 (持续大餐)
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as PredictionResult

        if (!fullInput.contentEquals(other.fullInput)) return false
        if (!noPatientInfo.contentEquals(other.noPatientInfo)) return false
        if (!lowCalorieMeal.contentEquals(other.lowCalorieMeal)) return false
        if (!mediumCalorieMeal.contentEquals(other.mediumCalorieMeal)) return false
        if (!highCalorieMeal.contentEquals(other.highCalorieMeal)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = fullInput.contentHashCode()
        result = 31 * result + noPatientInfo.contentHashCode()
        result = 31 * result + lowCalorieMeal.contentHashCode()
        result = 31 * result + mediumCalorieMeal.contentHashCode()
        result = 31 * result + highCalorieMeal.contentHashCode()
        return result
    }
}

/**
 * 患者数据类
 */
data class PatientData(
    val patientId: String,
    val historicalGlucose: FloatArray,  // 历史血糖值 (9个点)
    val timeSeriesFeatures: FloatArray, // 时序特征 (10 x 51 = 510)
    val staticFeatures: FloatArray       // 静态特征 (30)
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as PatientData

        if (patientId != other.patientId) return false
        if (!historicalGlucose.contentEquals(other.historicalGlucose)) return false
        if (!timeSeriesFeatures.contentEquals(other.timeSeriesFeatures)) return false
        if (!staticFeatures.contentEquals(other.staticFeatures)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = patientId.hashCode()
        result = 31 * result + historicalGlucose.contentHashCode()
        result = 31 * result + timeSeriesFeatures.contentHashCode()
        result = 31 * result + staticFeatures.contentHashCode()
        return result
    }
}
